// node-themoviedb/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var DataController_1 = __importDefault(require("./DataController"));
var accountEndpoint_1 = __importDefault(require("./endpoints/accountEndpoint"));
var authenticationEndpoint_1 = __importDefault(require("./endpoints/authenticationEndpoint"));
var certificationEndpoint_1 = __importDefault(require("./endpoints/certificationEndpoint"));
var collectionEndpoint_1 = __importDefault(require("./endpoints/collectionEndpoint"));
var companyEndpoint_1 = __importDefault(require("./endpoints/companyEndpoint"));
var configurationEndpoint_1 = __importDefault(require("./endpoints/configurationEndpoint"));
var creditEndpoint_1 = __importDefault(require("./endpoints/creditEndpoint"));
var discoverEndpoint_1 = __importDefault(require("./endpoints/discoverEndpoint"));
var findEndpoint_1 = __importDefault(require("./endpoints/findEndpoint"));
var genreEndpoint_1 = __importDefault(require("./endpoints/genreEndpoint"));
var guestSessionEndpoint_1 = __importDefault(require("./endpoints/guestSessionEndpoint"));
var keywordEndpoint_1 = __importDefault(require("./endpoints/keywordEndpoint"));
var listEndpoint_1 = __importDefault(require("./endpoints/listEndpoint"));
var movieEndpoint_1 = __importDefault(require("./endpoints/movieEndpoint"));
var networkEndpoint_1 = __importDefault(require("./endpoints/networkEndpoint"));
var trendingEndpoint_1 = __importDefault(require("./endpoints/trendingEndpoint"));
var personEndpoint_1 = __importDefault(require("./endpoints/personEndpoint"));
var reviewEndpoint_1 = __importDefault(require("./endpoints/reviewEndpoint"));
var searchEndpoint_1 = __importDefault(require("./endpoints/searchEndpoint"));
var tvEndpoint_1 = __importDefault(require("./endpoints/tvEndpoint"));
var MovieDB = /** @class */ (function () {
    function MovieDB(apiKey, options) {
        var _a;
        this.account = accountEndpoint_1.default;
        this.authentication = authenticationEndpoint_1.default;
        this.certification = certificationEndpoint_1.default;
        this.collection = collectionEndpoint_1.default;
        this.company = companyEndpoint_1.default;
        this.configuration = configurationEndpoint_1.default;
        this.credit = creditEndpoint_1.default;
        this.discover = discoverEndpoint_1.default;
        this.find = findEndpoint_1.default;
        this.genre = genreEndpoint_1.default;
        this.guestSession = guestSessionEndpoint_1.default;
        this.keyword = keywordEndpoint_1.default;
        this.list = listEndpoint_1.default;
        this.movie = movieEndpoint_1.default;
        this.network = networkEndpoint_1.default;
        this.trending = trendingEndpoint_1.default;
        this.person = personEndpoint_1.default;
        this.review = reviewEndpoint_1.default;
        this.search = searchEndpoint_1.default;
        this.tv = tvEndpoint_1.default;
        var data = DataController_1.default.getInstance();
        data.set('apiKey', apiKey);
        data.set('language', (_a = options === null || options === void 0 ? void 0 : options.language) !== null && _a !== void 0 ? _a : 'en-US');
    }
    // eslint-disable-next-line class-methods-use-this
    MovieDB.prototype.setApiKey = function (apiKey) {
        DataController_1.default.getInstance().set('apiKey', apiKey);
    };
    // eslint-disable-next-line class-methods-use-this
    MovieDB.prototype.setLanguage = function (language) {
        DataController_1.default.getInstance().set('language', language);
    };
    return MovieDB;
}());
exports.default = MovieDB;
// For CommonJS default export support
module.exports = MovieDB;
module.exports.default = MovieDB;
